command/agent/http.go

```diff
@@ -193,6 +194,58 @@ func NewHTTPServers(agent *Agent, config *Config) ([]*HTTPServer, error) {
 		srvs = append(srvs, srv)
 	}
 
+	// Initialize the API socket if configured
+	startAPISocket := func() {
+		if config.APISocket.Path == "" {
+			return
+		}
+
+		const (
+			MaxSocketPathLength = 104
+			msgPrefix           = "error starting api socket:"
+		)
+
+		if len(config.APISocket.Path) > MaxSocketPathLength {
+			serverInitializationErrors = multierror.Append(serverInitializationErrors,
+				fmt.Errorf("%s path must be %v characters or less", msgPrefix, MaxSocketPathLength))
+		}
+
+		sl, err := listenerutil.UnixSocketListener(config.APISocket.Path, config.APISocket.UnixSocketsConfig())
+		if err != nil {
+			serverInitializationErrors = multierror.Append(serverInitializationErrors,
+				fmt.Errorf("%s %v", msgPrefix, err))
+			return
+		}
+
+		// Create the server
+		srv := &HTTPServer{
+			agent:        agent,
+			eventAuditor: agent.auditor,
+			mux:          http.NewServeMux(),
+			listener:     sl,
+			listenerCh:   make(chan struct{}),
+			logger:       agent.httpLogger,
+			Addr:         sl.Addr().String(),
+			wsUpgrader:   wsUpgrader,
+		}
+		srv.registerHandlers(config.EnableDebug)
+
+		// Create HTTP server with timeouts
+		httpServer := http.Server{
+			Addr:      srv.Addr,
+			Handler:   handlers.CompressHandler(srv.mux),
```

Comment:
Since the unix socket lacks mTLS, I think we need to require authentication. We have a few HTTP endpoints that are unauthenticated because we consider mTLS a "good enough" baseline of access control.

It could be argued that the unix file permissions play the role of mTLS here, but (1) UDS permissions are a mess of OS-specific details and (2) now we're adding yet-another-dimension to our already highly complex auth story.

There are some options here:
1. Require auth only if ACLs are enabled. I think this would make the most sense to users and be met with little friction.
2. Require auth only if TLS (and ACLs?) are enabled. That actually matches the security properties we're going after, but I feel like it's going to get confusing for users.
3. Require auth but add a `api_unix_socket.allow_unauthenticated` option that **defaults to `false`.** The nice thing about this is that it gives us somewhere to drop the big long explanation of why we require it in the first place.

Thoughts? I really don't want to diminish the usefulness of this feature, but I also don't want to turn around have even a low sev CVE because we subtly changed our security model.

Reply:
I'm not sure it's all that bad? Most HTTP API calls get mapped to RPCs and those RPCs happen between agents over a separate TCP connection, so the HTTP API's mTLS doesn't ever get consulted as part of all that complex auth logic. The only HTTP APIs that don't get mapped to RPCs are those that hit the local agent anyways. So the HTTP API's mTLS is only used for establishing the connection between the API client and the HTTP server.

So treating the UDS permissions as the equivalent "transport security" for a host-local connection seems ok to me. Unless the UDS socket's permissions are wide-open, anyone who can talk to the UDS can probably grab the TLS cert anyways and spoof that.

That being said, I also suspect a lot of folks run with ACLs disabled on clients, because it's the default and most features work fine without it. Defaulting to requiring ACLs here could encourage wider use of client ACLs and give us a better path to always requiring ACLs (similar to our discussion around the Task API).

Reply:
**tl;dr - Ignore me. This is fine as-is.** Feel free to mark resolved.

> So treating the UDS permissions as the equivalent "transport security" for a host-local connection seems ok to me. Unless the UDS socket's permissions are wide-open, anyone who can talk to the UDS can probably grab the TLS cert anyways and spoof that.

Yeah, this makes sense. I think I was ignoring that fact that *for users on Client hosts unix permissions are already the de facto "transport" security.* It's just right now it's unix permissions on the *mTLS client certificates,* not a UDS file. So while adding a UDS is adding one extra file to a an Agent's security perimeter, it doesn't introduce any new concepts, implicit environment requirements, or complexity that mTLS didn't already.

Also I keep forgetting that we shouldn't even really worry about the `acl.enabled = false` configuration as it's outside our security model (despite being the default configuration).